### ISSUE TITLE
Lazily process installed packages

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#17924]
## Overview

Currently if you import anything from a third-party package that has inline types, mypy will process all the transitive dependencies of the imported module as well. This can be pretty slow, since some packages have hundreds or thousands of dependencies (e.g. `torch`, but there are probably many others).

We could speed this up by only processing those dependencies that are needed to type check code that uses the third-party package. This is possible, since we won't generally report errors from installed packages. This wouldn't be possible for normal code, since we could have false negatives in code that we don't process. We'd process (some) imported definitions lazily.

## Example

Assume we have third-party package `acme` that has 1000s of recursive module dependencies. Now we have user code that only uses one function from the top-level module:
```py
from acme import do_stuff
do_stuff()
```
We might only need to process `acme/__init__.py` to type check this code. Most of the 1000 dependencies can be ignored, and they don't even need to be parsed. However, if `do_stuff` or other functions in `acme/__init__.py` use a type in an annotation that is defined in a submodule of `acme`, we might need to process modules that define those types as well, and any dependencies they might have. (This assumes module-level granularity of laziness. It's easy to also imagine definition-level laziness, so that only the `do_stuff` function would have to be processed.)

## Implementation sketch

Here's a sketch of a potential implementation:
 * Add a flag for packages where we don't report errors in any recursive dependencies. This should be enabled for installed packages and stubs.
 * Make sure installed packages and stubs can't import code where we do report errors. They should only import stubs and other installed packages. Otherwise there would be false negatives. 
 * When processing an import targeting a "recursive no-error" package, initially don't consider any module dependencies. Add import placeholders to symbol table for imported symbols (unless they are already available).
 * If any code uses an import placeholder, defer the current node and keep track of the name of the placeholder target name. If any deferrals are due to import placeholders, process the import placeholders first before reprocessing the deferred nodes.
 * An name that is available as a placeholder only would be "unresolved".

## Discussion

Discussion:

 * Imports that are only used within function bodies in installed packages don't need to be resolved. This could be a big win.
 * Imports of other modules/functions/classes within installed package that are not used don't need to be resolved. This could help with packages that have massive public APIs.
 * It would be easier if we can resolve all placeholders during semantic analysis. I'm not sure if this is possible, at least due to modules being able to implement protocols.
 * If we analyze a class, we should maybe resolve all import placeholders in the class and any base classes to avoid having to resolve them during type checking.
 * Also if we analyze a type annotation, we should maybe resolve all placeholders related to the type so that we can perform type checking without dealing with unresolved references.
 * As an optimization, we could do a quick AST pass to determine any imported names that are definitely needed based on a shallow syntactic analysis (e.g. look for `from acme import submodule`). When processing a module, we'd resolve these first to avoid numerous deferrals. I'm not sure if this would be a big perf win or not.
* Since big SCCs are common, we need to able to do this within SCCs, not just between SCCs.
  * Circular dependencies need some care. We already support them, and probably the existing approach could be generalized.

Modules used as protocol-typed values could be an issue, since this could require arbitrary attributes (including nested imported modules) to be available, and we'd only know about this during type checking. So we might need to defer during type checking, and within various type operations such as subtype checks. This is probably pretty rare but still currently supported. Since this is expected to be rare, this doesn't need to be super efficient.

**Comment by JukkaL** [comment]
Because of the issues outlined above, we should probably start with lazily deserializing cache files. This would be easier to implement, and it would have some other nice properties:
 * Incremental runs are likely more common than non-incremental runs (outside CI).
 * This would help with all modules, not just third-party dependencies.
 * This should play nicely with parallel type checking.

**Comment by JukkaL** [comment]
I've thought more about this, and there's another interesting approach that could work. I'm not sure which of the options discussed here is the best yet, but this might be among the easiest to implement if we replace our parser.

This assumes that we'd replace the Python `ast` module with a significantly faster parser that doesn't construct an intermediate Python AST before converting it to a mypy AST. Likely the parser would not be implemented in Python but in Rust, or possibly C. There are good reasons for replacing the parser anyway that I won't discuss here in detail. Another assumption is that the parser can quickly and easily parse subtrees of modules, and that it can produce a "shallow" AST where some subtrees are not parsed.

Now we could perform a quick initial parsing pass against all recursive dependencies, but only parse the top levels of modules, and leave "placeholders" for any functions. Functions would only be parsed and analyzed lazily when they are needed.

Discussion:
* By processing functions lazily and by having a faster parser, we could probably improve processing speed by 4x or more (for installed packages).
* Since the parser would not be implemented in Python, we could probably run it in parallel, even on non-free-threaded builds. This would further improve efficiency.
* The remaining work beyond "shallow" parsing would involve fully analyzing module top levels. It would probably make sense to profile this and further optimize it, since this sounds quite repetitive and thus easy to optimize.
* The cache files would also not include most functions but have these placeholders. Thus deserialization of cache files would also be lazy.
* Normal code (not in installed packages or stubs) would still be fully processed (this could change in the future if we'd introduce an LSP mode).
* Classes seem a bit tricky. We need to be able to perform variance inference lazily. We need function bodies to find initialized attributes. If a class is used in any non-trivial way, we might have to process all methods at that point, including any base class methods.
* We'll always have a full module graph, but some (non-module) nodes would be placeholders in symbol tables.
* The low-level parser probably has to figure out basic reachability, including reachable imports (both inside and outside functions). So it may need to parse everything, but we won't generate mypy ASTs for everything. This assumes parsing itself is very fast.
* As always, we can strip away most function bodies in installed packages. With the hypothetical new parser, we wouldn't even need to construct mypy AST nodes for most function bodies.

It might be possible to also parse module top levels lazily, if we can just find imports quickly. So we'd do an initial pass over a module, where we only find import dependencies and populate imported references to other modules. When something is actually used from the module (other than something that has bee trivially imported from another module), we'd process it fully, or at least process the top level fully. This sounds more difficult to implement and has more open questions than the "lazy function" approach, but it could be even faster if feasible.

**Comment by JukkaL** [comment]
@ilevkivskyi recently contributed as new binary cache format, with fully custom deserialization and serialization code. Beyond massively improving deserialization speed, this seems to enable a simple implementation of partial lazy deserialization of cache files. Here's a summary of the idea:
* We'd still deserialize all module dependencies, but we'd deserialize the contents of each cached module lazily.
* During the initial deserialization of a module, function and variable nodes in symbol table would be deserialized as placeholder objects that contain the serialized byte string for the function/variable. This would cover a large fraction of deserialized nodes.
* We'd allocate the actual AST nodes (e.g. `FuncDef`) when the symbol table node is first accessed after the initial deserialization pass, e.g. during type checking. My hypothesis is that most methods and top-level functions in third-party libraries don't usually need to be deserialized, as they are never accessed outside the module that defines them.

At least one part of the implementation is a bit tricky -- when deserializing symbol tables, we'd need to perform "shallow deserialization" (or partial/lazy deserialization). I think we can implement this efficiently if we have a way to quickly find the end of the serialized data of an arbitrary object. The approach I proposed in #19697 seems to enable this, but there are other ways to do this as well.

This seems like a fairly easy win now that @ilevkivskyi did the heavy lifting with the new serialization format, and I might try to implement this in the near future.

**Comment by ilevkivskyi** [comment]
We discussed with @JukkaL offline and it seems like adding type tags would be a future-proof thing to do. Not  just for lazy deserialization, but in general. So I think we should use them to quickly find the end of serialized data for a symbol.
